core/drivers/atmel_shdwc.c

```diff
@@ -0,0 +1,194 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (c) 2015 Atmel Corporation,
+ *                    Nicolas Ferre <nicolas.ferre@atmel.com>
+ * Copyright (c) 2021, Microchip
+ */
+
+#include <at91_clk.h>
+#include <drivers/atmel_shdwc.h>
+#include <drivers/sam/at91_ddr.h>
+#include <io.h>
+#include <kernel/dt.h>
+#include <libfdt.h>
+#include <stdbool.h>
+#include <tee_api_defines.h>
+#include <tee_api_types.h>
+#include <trace.h>
+#include <types_ext.h>
+#include <util.h>
+
+#define SHDW_WK_PIN(reg, cfg)	((reg) & \
+					AT91_SHDW_WKUPIS((cfg)->wkup_pin_input))
+#define SHDW_RTCWK(reg, cfg)	(((reg) >> ((cfg)->sr_rtcwk_shift)) & 0x1)
+#define SHDW_RTTWK(reg, cfg)	(((reg) >> ((cfg)->sr_rttwk_shift)) & 0x1)
+#define SHDW_RTCWKEN(cfg)	BIT((cfg)->mr_rtcwk_shift)
+#define SHDW_RTTWKEN(cfg)	BIT((cfg)->mr_rttwk_shift)
+
+#define SLOW_CLK_FREQ		32768ULL
+#define DBC_PERIOD_US(x)	DIV_ROUND_UP((1000000ULL * (x)), SLOW_CLK_FREQ)
+
+static vaddr_t shdwc_base;
+static vaddr_t mpddrc_base;
+
+bool atmel_shdwc_available(void)
+{
+	return shdwc_base != 0;
+}
+
+void __noreturn atmel_shdwc_shutdown(void)
+{
+	vaddr_t pmc_base = at91_pmc_get_base();
+
+	asm volatile (
```

Comment:
Ok, best solution for this is to move it in assembly file. The output must be controlled to avoid being larger than a cache line since we power down the SDRAM.

Reply:
Sounds good. Would you mind adding a comment in the assembly file to explain why it has to be written in assembly?

Reply:
Sure !